packages/shared/src/utils/deprecate.ts

```diff
@@ -0,0 +1,87 @@
+type AnyFn = (...args: any[]) => any;
+type AnyClass = new (...args: any[]) => any;
+
+type DeprecateType = (
+  fnOrPropName: string,
+  warning: string,
+  fnOrMethodOrCls: AnyFn | AnyClass,
+  isStatic?: boolean,
+) => AnyFn;
+
+/**
+ * Mark class properties/methods or functions as deprecated.
+ *
+ * On marked function/methods/class props a console.warn
+ * will be shown ONLY for non-production environments on the
+ * first time the deprecated object will be accessed or invoked.
+ *
+ * Examples
+ * 1. Deprecate class method
+ * class Example {
+ *   getSomething = deprecate('getSomething', 'Use `getSomethingElse` instead.', (arg1, arg2) => {
+ *       return `getSomethingValue:${arg1 || '-'}:${arg2 || '-'}`;
+ *     });
+ *   }
+ *
+ * 2. Deprecate class property
+ * class Example {
+ *   something: string;
+ *   constructor(something: string) {
+ *     this.something = something;
+ *   }
+ * }
+ *
+ * deprecate('something', 'Use `somethingElse` instead.', Example);
+ *
+ * 3. Deprecate class static property
+ * class Example {
+ *   static something: string;
+ * }
+ *
+ * deprecate('something', 'Use `somethingElse` instead.', Example, true);
+ *
+ * 4. Deprecate function
+ * const getSomething = deprecate('getSomething', 'Use `getSomethingElse` instead.', () => {
+ *   return 'getSomethingValue';
+ * });
+ */
+export const deprecate: DeprecateType = (fnOrPropName, warning, fnOrMethodOrCls, isStatic = false) => {
+  let hasWarned = false;
+
+  const warn = () => {
+    const isProduction = process.env.NODE_ENV === 'production';
+
+    if (isProduction) {
+      return;
+    }
```

Comment:
❓ Should we go the other way around for this? `Do not show warning if it's not in development mode.`